 Apps on adopted storage should stay there during update

An update to an app on adopted storage via a PackageInstaller.Session
(eg gplay) will result in the app silently ending up back on internal
storage if the (new) manifest sets INSTALL_LOCATION_INTERNAL_ONLY.
However, the app data is not moved with it so the end result is that
data looks like it's been wiped.

Ideally, the app and data would be moved to internal.  Next best would
be that the user is informed that they need to uninstall and reinstall.

For now, persist the app in the adopted storage location so that
the user doesn't lose app data.



diff --git a/core/java/com/android/internal/content/PackageHelper.java b/core/java/com/android/internal/content/PackageHelper.java
index b04ddf4..0611401 100644
--- a/core/java/com/android/internal/content/PackageHelper.java
+++ b/core/java/com/android/internal/content/PackageHelper.java
@@ -378,15 +378,6 @@
             installLocation = PackageInfo.INSTALL_LOCATION_INTERNAL_ONLY;
         }
 
-        // If app expresses strong desire for internal space, honor it
-        if (installLocation == PackageInfo.INSTALL_LOCATION_INTERNAL_ONLY) {
-            if (fitsOnInternal) {
-                return null;
-            } else {
-                throw new IOException("Requested internal only, but not enough space");
-            }
-        }
-
         // If app already exists somewhere, prefer to stay on that volume
         if (existingInfo != null) {
             if (existingInfo.volumeUuid == null && fitsOnInternal) {
@@ -397,6 +388,15 @@
             }
         }
 
+        // If app expresses strong desire for internal space, honor it
+        if (installLocation == PackageInfo.INSTALL_LOCATION_INTERNAL_ONLY) {
+            if (fitsOnInternal) {
+                return null;
+            } else {
+                throw new IOException("Requested internal only, but not enough space");
+            }
+        }
+
         // We're left with either preferring external or auto, so just pick
         // volume with most space
         if (bestCandidate != null) {
